File: Ruby/napakalaki/lib/cultist_player.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module NapakalakiGame

require_relative 'treasure'
require_relative 'monster'
require_relative 'treasure_kind'
require_relative 'combat_result'
require_relative 'dice'
require_relative 'napakalaki'
require_relative 'player'

class CultistPlayer < Player
  
  @@totalCultistsPlayers = 0
  
  def initialize(player, cultist)
    
    @@totalCultistsPlayers += 1
   
    copy(player)
   
    @cultistCard = cultist
    
  end
  
  def getCombatLevel
    
    (super * 0.7).to_i + (@cultistCard.getGainedBonus * @@totalCultistsPlayers)
    
  end
  
  def shouldConvert
    false
  end
  
  def getOponentLevel(m)
    m.getCombatLevelAgainstCultist
  end
  
  def giveMeATreasure
    
    t = @visibleTreasures[rand(@visibleTreasures.size)]
    discardVisibleTreasure(t)
    return t
    
  end
  
  def canYouGiveMeATreasure
    @visibleTreasures.any?
  end
  
  def to_s
    "#{@name} (Sectario) (Nivel de jugador: #{@level}  Nivel de combate: #{getCombatLevel}  Enemigo: #{@enemy.getName})"
  end
  
end
end

File: Java/src/napakalaki/CultistPlayer.java

package napakalaki;

public class CultistPlayer extends Player {
    
    private static int totalCultistsPlayers = 0;
    private final Cultist myCultistCard;
    
    public CultistPlayer(Player p, Cultist c){
        
        super(p);
        myCultistCard = c;
        CultistPlayer.totalCultistsPlayers++;

    }
    
    //CombatLevel of a cultist is increased.
    @Override
    public int getCombatLevel(){
        return (int)(1.7 * super.getCombatLevel()) + this.myCultistCard.getGainedBonus()*CultistPlayer.totalCultistsPlayers;
    }
    
    //A cultist can't convert again.
    @Override
    public boolean shouldConvert(){
        return false;
    }
    
    //Monster might have a different combat level versus a cultist.
    @Override 
    protected int getOponentLevel(Monster m){
        return m.getCombatLevelAgainstCultistPlayer();
    }
    
    //When stolen they lose a visible treasure.
    @Override
    protected Treasure giveMeATreasure(){
        
        return this.deleteAndReturnRandomIn(this.visibleTreasures);
    }

    @Override
    protected boolean canYouGiveMeATreasure(){
        return !this.visibleTreasures.isEmpty();
    }
    
    @Override
    public String toString(){
        return super.toString() + "(Sectario)";
    }
    
    public static int getTotalCultistsPlayers(){
        return totalCultistsPlayers;
    }
  
}
